python-package/xgboost/data.py

```diff
@@ -317,7 +317,7 @@ def _is_cudf_df(data):
         import cudf
     except ImportError:
         return False
-    return isinstance(data, cudf.DataFrame)
+    return hasattr(cudf, 'DataFrame') and isinstance(data, cudf.DataFrame)
```

Comment:
?

Reply:
When the in-place predictor calls `_is_cudf_df` from multiple threads, some threads get a stub `cudf` module which nevertheless does not contain `DataFrame`. You can try it yourself with CUDA 11 container.

Reply:
That's strange, is it an issue on our side?  Who created the stub?  I'm fine with the change, just trying to understand so I know what to do if I run into it.

Reply:
More precisely, in at least one thread, the `ImportError` exception is not thrown for some reason.

Maybe we should add `@pytest.mark.skipif(**tm.no_cudf())` to the cuPy in-place predictor test?

Reply:
Oh no,  this function got used in inplace predict.  I should remove it as it creates lock in Python.

> , in at least one thread, the ImportError exception is not thrown for some reason.

Probably a Python bug?

> Maybe we should add @pytest.mark.skipif(**tm.no_cudf()) to the cuPy in-place predictor test?

No, your patch looks good.  It helps real use case instead of test.